js-packages/@prestojs/viewmodel/src/ViewModelFactory.ts

```diff
@@ -157,13 +152,42 @@ export interface ViewModelConstructor<
     __pkFieldType: PkFieldType;
     /** @private */
     __pkType: PkType;
-    new <
-        IncomingData extends FieldDataMappingRaw<FieldMappingType> = FieldDataMappingRaw<
-            FieldMappingType
-        >
-    >(
-        data: IncomingData
-    ): ViewModelInterface<FieldMappingType, IncomingData, PkFieldType, PkType>;
+    new <FieldNames extends keyof FieldMappingType>(
+        data: { [K in FieldNames]: FieldMappingType[K]['__parsableValueType'] }
+    ): ViewModelInterface<FieldMappingType, FieldNames, PkFieldType, PkType>;
+
+    // This is necessary for when the base class is extended so that it still conforms to
+    // ViewModelConstructor - without this there will be no matching type (for some reason?)
+    // due to `IncomingData`. It's important to fix this otherwise anywhere that accepts
+    // ViewModelConstructor will fail to accept anything that extends a class created with
+    // viewModelFactory by default.
+    //
+    // I don't fully understand why. This means that for a class
+    // that extends the base class you won't get proper types for partial records. For example:
+    // const A = viewModelFactory({
+    //     name: new Field<string>(),
+    //     age: new Field<number>(),
+    // });
+    //
+    // const a1 = new A({ name: 'test' });
+    // // Error: Property 'age' does not exist on type ...
+    // a1.age;
+    // class B extends A {}
+    // const a2 = new B({ name: 'test' });
+    // // No error, all known fields are valid:
+    // a2.age
+    //
+    // You can fix it with:
+    // const C: ViewModelConstructor<any> = B;
+    // const a3 = new C({ name: 'test' });
+    // Error: Property 'age' does not exist on type ...
+    // a3.age
+    new (data: FieldDataMappingRaw<FieldMappingType>): ViewModelInterface<
+        FieldMappingType,
+        keyof FieldMappingType,
+        PkFieldType,
+        PkType
+    >;
```

Comment:
is this what we discussed here? https://alliance-software.slack.com/archives/CGPEMBTJL/p1587974607004400 :p

Reply:
yeah basically that